samples/isprime/isprime.neon

```diff
@@ -0,0 +1,184 @@
+   in this sequence and is immediately preceded by -1. If the test 
+   fails, we know that n is composite, and describe the base a as 
+   a witness to the compositeness of n. The crucial extra information 
+   we can derive is that if (odd) n is composite, then at least 3/4 
+   of the numbers a with 1 < a < n - 1 are witness to this fact.
+   First calculate m, in Neon we do this by simply shifting the bits
+   right k times, until m is odd and less than n.
+   Next, we calculate a, such that a is less than n - 1, and a > 1.
+   Then, we calculate a ^ m mod n.  If this result is 1, then
+   we need to generate a new "a" and try again.  After iIterations, we
+   can say that n is prime with 1/1024 margin of error.
+ |%
+IMPORT binary
+IMPORT random
+IMPORT string
+IMPORT sys
+
+LET SMALL_PRIME: Array<Number> := [  2,   3,   5,   7,  11,  13,  17,  19,  23,  29,  31,  37,  41,  43,  47,  53,  59,  61, 
+                                    67,  71,  73,  79,  83,  89,  97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 
+                                   157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257 ]
+
+FUNCTION SizeOf(obj: Array<Number>): Number
+    RETURN obj.size()
+END FUNCTION
+                            
+CONSTANT TestIterations: Number := 5
+
+FUNCTION GetRandomA(OUT a:Number, n:Number)
+    VAR rnd: Number := random.uint32()
+    a := rnd MOD (n - 2) + 1
+END FUNCTION
+
+FUNCTION isOdd(n: Number): Boolean
+    IF binary.get32(n, 0) = TRUE THEN
+        RETURN TRUE
+    END IF
+    RETURN FALSE
+END FUNCTION
+
+FUNCTION RabinMiller(n: Number, iIterations: Number): Boolean
+    VAR isprime: Boolean:= FALSE
+    VAR m:       Number := n - 1
+    VAR k:       Number := 0
+    VAR a:       Number := 0
+    VAR i:       Number := iIterations
+
+    IF NOT isOdd(n) THEN
+        % Rule: The number n is an EVEN number, and therefore CANNOT be prime.
+        IF n = 2 THEN
+            % The number 2 is the only exception to the rule!
+            % Since 2 is the ONLY even number that is prime, we handle this
+            % case a little differently.
+            RETURN TRUE
+        END IF
+        RETURN isprime
+    END IF
+
+    IF m > 0 THEN
+        WHILE binary.get32(m, 0) = TRUE DO
+            INC k
+            m := binary.shiftRight32(m, 1)
+            IF m < 0 THEN 
+                EXIT WHILE
+            END IF
+        END WHILE
+    END IF
+
+    VAR r: Number
+    REPEAT
+        GetRandomA(OUT a, n)
+        r := 1
+        FOR x := 0 TO m - 1 DO
+            r := (r * a) MOD n
+        END FOR
+        IF r = 1 OR r = (n - 1) THEN
+            isprime := TRUE
+            DEC i
+        END IF
+        FOR x := 0 TO k DO
+            r := (r*r) MOD n
+            IF r = (n - 1) THEN
+                isprime := TRUE
+                EXIT FOR
+            END IF
+        END FOR
+        IF NOT isprime THEN
```

Comment:
Where would `isprime` ever get set to false in this loop? It looks like this whole loop tests the first iteration, setting isprime to true if it passes that first test, then keeps running the rest of the iterations which cannot possibly have any effect on the ultimate outcome.